src/recovery.rs

```diff
@@ -513,1439 +834,195 @@ fn migrations_valid() -> rusqlite_migration::Result<()> {
-                                    let key = (step_id, state_key);
-
-                                    (key, ser_snap)
-                                });
-                        ser_snaps_output.session(&cap).give_iterator(ser_snaps);
-                    });
-                });
-            }
-        })
-    }
-}
+                input.buffer_notify(&mut inbuffer.borrow_mut(), &mut ncater);
 
-trait DeserializeSnapshotOp<S>
-where
-    S: Scope,
-{
-    /// Deserialize state snapshots using the provided serde.
-    fn de_snap(&self) -> Stream<S, Snapshot>;
-}
+                ncater.for_each(
+                    input_frontiers,
+                    |caps, ()| {
+                        let cap = &caps[0];
 
-impl<S> DeserializeSnapshotOp<S> for Stream<S, SerializedSnapshot>
-where
-    S: Scope,
-{
-    fn de_snap(&self) -> Stream<S, Snapshot> {
-        self.map(
-            move |SerializedSnapshot(step_id, state_key, _snap_epoch, ser_change)| {
-                let snap_change = match ser_change {
-                    Some(ser_snap) => {
-                        let snap = unwrap_any!(Python::with_gil(|py| -> PyResult<PyObject> {
-                            let pickle = py.import_bound("pickle")?;
-                            Ok(pickle
-                                .call_method1(
-                                    intern!(py, "loads"),
-                                    (PyBytes::new_bound(py, &ser_snap),),
-                                )?
-                                .unbind())
-                        }));
-                        StateChange::Upsert(snap.into())
-                    }
-                    None => StateChange::Discard,
-                };
+                        let mut handle = immediate_segments_output.activate();
+                        let mut session = handle.session(cap);
 
-                Snapshot(step_id, state_key, snap_change)
-            },
-        )
-    }
-}
+                        let epochs = inbuffer.borrow().epochs().collect::<Vec<_>>();
 
-pub(crate) trait LoadSnapsOp<S>
-where
-    S: Scope<Timestamp = u64>,
-{
-    /// Read state data from the recovery partitions into the production dataflow.
-    ///
-    /// You will still need to route it to the correct steps. See
-    /// [`FilterSnapsOp::filter_snaps`].
-    ///
-    /// This will dump all state in the epoch of the snapshot so that
-    /// operators can load in epoch order.
-    fn load_snaps(&mut self, before: ResumeEpoch, bundle: RecoveryBundle) -> Stream<S, Snapshot>
-    where
-        S: Scope;
-}
+                        for epoch in epochs {
+                            let snaps = inbuffer.borrow_mut().remove(&epoch).unwrap();
+                            let path = local_state_store
+                                .borrow_mut()
+                                .write_snapshots_segment(snaps, epoch)
+                                .unwrap();
+                            session.give(path);
```

Comment:
This is probably the root problem that causes the things you saw in the other comments.
I'll first try to explain what I'm doing, but yes this is something that I missed while working on the fixes from the previous round of review, my bad.
The reason the `LocalStateStore` is shared between the operators, is that we need(ed) to keep track of the current segment number, which had to be known by both operators (now that we use a different naming for frontier and snpashot segments and a single frontier segment, that is not needed anymore). So rather then dropping the store itself, when I emit the path I change the value of the segment number, so that every operator will open a connection to the correct file. The issue is that segments are emitted as they change, rather than at the end of the epoch after a compaction, so this operator is not actually doing the compaction (I probably should have noticed this). I think the fix should not be too complicated at this point though. Working on this

Reply:
Each operator won't be able to open an already opened segment, given that:
- Each worker has its own separate file
- The connection is not persistent, it's opened/closed during each activation
- Operators on the same worker don't run concurrently

So that shouldn't be a problem of the current implementation, but I get your point about treating the path as an ownership token, and it still makes sense to make the changes since we don't need to share the segment number between operators anymore.

Reply:
Yes, that's clear, but that was ensured by the fact that each time an operator writes to a segment, it writes to a different one. If you see the code for `write_snapshots_segment`, the first thing it does is to change either the segment number alone, or both epoch number and segment number, changing the file used.